packages/shared/src/hooks/useBookmarkPost.ts

```diff
@@ -1,82 +1,218 @@
     },
   );
 
-  const { mutateAsync: removeBookmark } = useMutation<
-    void,
-    unknown,
-    T,
-    (() => void) | undefined
-  >(
-    ({ id }) =>
-      request(graphqlUrl, REMOVE_BOOKMARK_MUTATION, {
+  const addBookmark = useCallback(
+    ({ id }: BookmarkProps) => {
+      return bookmarkPost({
         id,
-      }),
-    {
-      onMutate: onRemoveBookmarkMutate,
-      onError: (err, _, rollback) => rollback?.(),
-      onSuccess: () =>
-        onRemoveBookmarkTrackObject &&
-        trackEvent(onRemoveBookmarkTrackObject()),
+        mutation: ADD_BOOKMARKS_MUTATION,
+        payload: { data: { postIds: [id] } },
+      });
     },
+    [bookmarkPost],
   );
-  const bookmarkToast = (targetBookmarkState) =>
-    targetBookmarkState
-      ? displayToast('Post was added to your bookmarks')
-      : displayToast('Post was removed from your bookmarks');
 
-  return {
-    bookmark,
-    bookmarkToast,
-    removeBookmark,
+  const removeBookmark = useCallback(
+    ({ id }: BookmarkProps) => {
+      return bookmarkPost({
+        id,
+        mutation: REMOVE_BOOKMARK_MUTATION,
+        payload: { id },
+      });
+    },
+    [bookmarkPost],
+  );
+
+  const toggleBookmark = useCallback(
+    async ({ post, origin, opts }: ToggleBookmarkProps) => {
+      if (!post) {
+        return;
+      }
+
+      if (!user) {
+        showLogin({ trigger: AuthTriggers.Bookmark });
+        return;
+      }
+
+      const analyticsOptions = prepareBookmarkPostAnalyticsOptions({
+        post,
+        origin,
+        opts,
+      });
+
+      if (post.bookmarked) {
+        trackEvent(
+          postAnalyticsEvent(
+            AnalyticsEvent.RemovePostBookmark,
+            post,
+            analyticsOptions,
+          ),
+        );
+        await removeBookmark({ id: post.id });
+        displayToast('Post was removed from your bookmarks');
+        return;
+      }
+
+      trackEvent(
+        postAnalyticsEvent(AnalyticsEvent.BookmarkPost, post, analyticsOptions),
+      );
+
+      await addBookmark({ id: post.id });
+      displayToast('Post was added to your bookmarks');
+    },
+    [addBookmark, displayToast, removeBookmark, showLogin, trackEvent, user],
+  );
+
+  return { toggleBookmark };
+};
+
+export { useBookmarkPost };
+
+type MutateBookmarkFeedPostProps = {
+  id: string;
+  updatePost: UpdateFeedPost;
+  items: FeedItem[];
+};
+
+export const mutateBookmarkFeedPost = ({
+  id,
+  items,
+  updatePost,
+}: MutateBookmarkFeedPostProps): ReturnType<
+  UseBookmarkPostProps['onMutate']
+> => {
+  if (!items) {
+    return undefined;
+  }
+
+  const postIndexToUpdate = items.findIndex(
+    (item) => item.type === 'post' && item.post.id === id,
+  );
+
+  if (postIndexToUpdate === -1) {
+    return undefined;
+  }
+
+  const mutationHandler = (post: Post) => ({ bookmarked: !post.bookmarked });
+  const previousState = (items[postIndexToUpdate] as PostItem)?.post
+    ?.bookmarked;
+
+  optimisticPostUpdateInFeed(
```

Comment:
I think it might be neater to sync the state with `useMutationSubscription` like `useFeedVotePost` so we don't have to bundle mutation of single and feed into single logic.

Reply:
Once we get a lean feed, we should be able to slim this down to just use one mutation using a single cache key for both feed and single post.

Reply:
Lean feed is not doing that currently. It is basically doing something similar to this by updating all the caches (at least currently). It is gonna take another iteration I think revamp the caching structure. @rebelchris you know more on how it is going.

Reply:
Correct first iteration will not fix that.
I think we could use the useMutation thing for now, it should handle logic a bit better right (more cleaner approach)